torch/csrc/jit/script/compiler.cpp

```diff
@@ -1253,6 +1247,13 @@ struct to_ir {
               << " in the false branch";
         if (save_true->findInParentFrame(x) ||
             save_false->findInParentFrame(x)) {
+
+          for(const auto& e : definition_errors) {
+            ErrorReport report(stmt);
```

Comment:
It looks like you're setting all definition errors for each variable that we iterate through

Reply:
I'm doing it right before we throw, so it should only happen once?

Reply:
ooooh, i see what is happening. 
`setVariableTypeError` is for future use, it still needs to be called even if we do not error here. 
```
if cond:
  i = 1
print(i)
```
we don't error at the point of emitting the if, but when `i` is accessed later we can print a more informative error message of why it is not defined

Reply:
Can you suggest another approach then?  Building these error strings eagerly is meaningfully expensive.

Reply:
You should be able to store all the information necessary to reconstruct the error message instead of the error message itself, and then construct the error message when it is thrown in `findVariableTypeError`.
For the case that the variable is only defined in one branch:
    - which branch it was not defined in (true/false), range of the if statement

For the case where there is a type mismatch so we do not emit the variable:
   - type of the True Branch, type of the False branch, range of the if statement. 

Or save the error message construct as a closure but i don't think that would be very helpfully.